File: src/stores/javaProjectStore.ts
import { create } from 'zustand';
import * as api from '../services/api';
import type { StoredJavaProject, ParsedJavaProject } from '../services/api';

export const buildEndpointCompositeKey = (controllerName: string, method: string, fullPath: string) =>
  `${controllerName}:${method}:${fullPath}`;

export const getEndpointCompositeKeys = (parsedData: ParsedJavaProject, endpointIds?: string[]) => {
  const endpointIdSet = endpointIds ? new Set(endpointIds) : null;

  return parsedData.controllers.flatMap((controller) =>
    controller.endpoints
      .filter((endpoint) => !endpointIdSet || endpointIdSet.has(endpoint.id))
      .map((endpoint) => buildEndpointCompositeKey(controller.name, endpoint.httpMethod, endpoint.fullPath)),
  );
};

export const getNewEndpointIds = (parsedData: ParsedJavaProject, seenEndpointIds: string[]) => {
  const seenSet = new Set(seenEndpointIds);

  return parsedData.controllers.flatMap((controller) =>
    controller.endpoints
      .filter((endpoint) => !seenSet.has(buildEndpointCompositeKey(controller.name, endpoint.httpMethod, endpoint.fullPath)))
      .map((endpoint) => endpoint.id),
  );
};

interface JavaProjectState {
  // State
  projects: StoredJavaProject[];
  currentProject: StoredJavaProject | null;
  parsedData: ParsedJavaProject | null;
  newEndpointIds: string[];
  importedEndpointIds: string[];
  isLoading: boolean;
  isLoaded: boolean;
  error: string | null;

  // Actions
  loadProjects: () => Promise<void>;
  addProject: (name: string, path: string, parsedData: ParsedJavaProject) => Promise<StoredJavaProject>;
  openProject: (projectId: string) => Promise<void>;
  closeProject: (projectId: string) => Promise<void>;
  deleteProject: (projectId: string) => Promise<void>;
  checkForUpdates: (projectId: string) => Promise<void>;
  markEndpointsSeen: (projectId: string, endpointIds: string[]) => Promise<void>;
  markImportedEndpoints: (endpointIds: string[]) => void;
  resetImportedEndpoints: () => void;
  setParsedData: (data: ParsedJavaProject | null) => void;
  setCurrentProject: (project: StoredJavaProject | null) => void;
}

export const useJavaProjectStore = create<JavaProjectState>((set, get) => ({
  // Initial state
  projects: [],
  currentProject: null,
  parsedData: null,
  newEndpointIds: [],
  importedEndpointIds: [],
  isLoading: false,
  isLoaded: false,
  error: null,

  // Actions
  loadProjects: async () => {
    const { isLoaded, isLoading } = get();
    if (isLoading || isLoaded) {
      return;
    }

    set({ isLoading: true, error: null });
    try {
      const storage = await api.getJavaProjects();
      set({ projects: storage.projects, isLoading: false, isLoaded: true });
    } catch (e) {
      set({ error: String(e), isLoading: false });
    }
  },

  addProject: async (name: string, path: string, parsedData: ParsedJavaProject) => {
    const now = new Date().toISOString();

    const seenEndpointIds = getEndpointCompositeKeys(parsedData);

    const project: StoredJavaProject = {
      id: crypto.randomUUID(),
      name,
      path,
      isOpen: true,
      lastParsedAt: now,
      seenEndpointIds,
    };

    await api.saveJavaProject(project);

    const { projects } = get();
    // Check if project with same path exists
    const existingIndex = projects.findIndex(p => p.path === path);
    if (existingIndex >= 0) {
      const updatedProjects = [...projects];
      updatedProjects[existingIndex] = project;
      set({ projects: updatedProjects, currentProject: project, parsedData, importedEndpointIds: [], isLoaded: true });

    } else {
      set({ projects: [...projects, project], currentProject: project, parsedData, importedEndpointIds: [], isLoaded: true });

    }

    return project;
  },

  openProject: async (projectId: string) => {
    set({ isLoading: true, error: null });
    try {
      await api.setJavaProjectOpen(projectId, true);

      const { projects } = get();
      const project = projects.find(p => p.id === projectId);
      if (project) {
        // Re-parse the project to get latest data
        const response = await api.checkJavaProjectUpdates(projectId);

        const updatedProject = { ...project, isOpen: true };
        const updatedProjects = projects.map(p => p.id === projectId ? updatedProject : p);

        set({
          projects: updatedProjects,
          currentProject: updatedProject,
          parsedData: response.parsedData,
          newEndpointIds: response.newEndpointIds,
          importedEndpointIds: [],
          isLoading: false,
          isLoaded: true,
        });
      }
    } catch (e) {
      set({ error: String(e), isLoading: false });
    }
  },

  closeProject: async (projectId: string) => {
    try {
      await api.setJavaProjectOpen(projectId, false);

      const { projects, currentProject } = get();
      const updatedProjects = projects.map(p =>
        p.id === projectId ? { ...p, isOpen: false } : p
      );

      set({
        projects: updatedProjects,
        currentProject: currentProject?.id === projectId ? null : currentProject,
        parsedData: currentProject?.id === projectId ? null : get().parsedData,
        newEndpointIds: [],
        importedEndpointIds: currentProject?.id === projectId ? [] : get().importedEndpointIds,
        isLoaded: true,
      });
    } catch (e) {
      set({ error: String(e) });
    }
  },

  deleteProject: async (projectId: string) => {
    try {
      await api.deleteJavaProject(projectId);

      const { projects, currentProject } = get();
      const updatedProjects = projects.filter(p => p.id !== projectId);

      set({
        projects: updatedProjects,
        currentProject: currentProject?.id === projectId ? null : currentProject,
        parsedData: currentProject?.id === projectId ? null : get().parsedData,
        newEndpointIds: [],
        importedEndpointIds: currentProject?.id === projectId ? [] : get().importedEndpointIds,
        isLoaded: true,
      });
    } catch (e) {
      set({ error: String(e) });
    }
  },

  checkForUpdates: async (projectId: string) => {
    try {
      const response = await api.checkJavaProjectUpdates(projectId);
      set({
        parsedData: response.parsedData,
        newEndpointIds: response.newEndpointIds,
        isLoaded: true,
      });
    } catch (e) {
      set({ error: String(e) });
    }
  },

  markEndpointsSeen: async (projectId: string, endpointIds: string[]) => {
    try {
      await api.markJavaEndpointsSeen(projectId, endpointIds);

      const { projects, currentProject, newEndpointIds } = get();

      // Update local state
      const updatedProjects = projects.map(p => {
        if (p.id === projectId) {
          return {
            ...p,
            seenEndpointIds: [...new Set([...p.seenEndpointIds, ...endpointIds])],
          };
        }
        return p;
      });

      // Remove marked IDs from newEndpointIds
      const remainingNewIds = newEndpointIds.filter(id => !endpointIds.includes(id));

      set({
        projects: updatedProjects,
        currentProject: currentProject?.id === projectId
          ? { ...currentProject, seenEndpointIds: [...new Set([...currentProject.seenEndpointIds, ...endpointIds])] }
          : currentProject,
        newEndpointIds: remainingNewIds,
        isLoaded: true,
      });
    } catch (e) {
      set({ error: String(e) });
    }
  },

  markImportedEndpoints: (endpointIds: string[]) => {
    const { importedEndpointIds } = get();
    set({ importedEndpointIds: [...new Set([...importedEndpointIds, ...endpointIds])] });
  },

  resetImportedEndpoints: () => {
    set({ importedEndpointIds: [] });
  },

  setParsedData: (data: ParsedJavaProject | null) => {
    set({ parsedData: data });
  },

  setCurrentProject: (project: StoredJavaProject | null) => {
    set({ currentProject: project });
  },
}));

File: src-tauri/src/storage/java_project_store.rs
use crate::commands::java_cmd::ParsedJavaProject;
use crate::storage::file_manager::FileManager;
use crate::utils::error::AppResult;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

const JAVA_PROJECTS_FILE: &str = "java_projects.json";

/// Stored Java project reference
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StoredJavaProject {
    pub id: String,
    pub name: String,
    pub path: String,
    pub is_open: bool,
    pub last_parsed_at: String,
    /// Store endpoint IDs that user has seen (for detecting new ones)
    pub seen_endpoint_ids: Vec<String>,
}

/// Java projects storage
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct JavaProjectsStorage {
    pub projects: Vec<StoredJavaProject>,
}

pub struct JavaProjectStore {
    file_manager: FileManager,
}

impl JavaProjectStore {
    pub fn new() -> Self {
        Self {
            file_manager: FileManager::new(),
        }
    }

    /// Get the storage file path
    fn get_storage_path(app_handle: &AppHandle) -> AppResult<PathBuf> {
        let app_data_dir = app_handle.path().app_data_dir()?;
        Ok(app_data_dir.join(JAVA_PROJECTS_FILE))
    }

    /// Load all stored Java projects
    pub async fn load_projects(&self, app_handle: &AppHandle) -> AppResult<JavaProjectsStorage> {
        let storage_path = Self::get_storage_path(app_handle)?;

        if !self.file_manager.exists(&storage_path).await {
            return Ok(JavaProjectsStorage::default());
        }

        self.file_manager.read_json(&storage_path).await
    }

    /// Save Java projects storage
    pub async fn save_projects(&self, app_handle: &AppHandle, storage: &JavaProjectsStorage) -> AppResult<()> {
        let storage_path = Self::get_storage_path(app_handle)?;

        // Ensure the directory exists
        if let Some(parent) = storage_path.parent() {
            self.file_manager.create_dir(parent).await?;
        }

        self.file_manager.write_json(&storage_path, storage).await
    }

    /// Add a new Java project
    pub async fn add_project(&self, app_handle: &AppHandle, project: StoredJavaProject) -> AppResult<()> {
        let mut storage = self.load_projects(app_handle).await?;

        // Check if project already exists (by path)
        if let Some(existing) = storage.projects.iter_mut().find(|p| p.path == project.path) {
            // Update existing project
            existing.name = project.name;
            existing.last_parsed_at = project.last_parsed_at;
            existing.seen_endpoint_ids = project.seen_endpoint_ids;
        } else {
            storage.projects.push(project);
        }

        self.save_projects(app_handle, &storage).await
    }

    /// Update a Java project
    pub async fn update_project(&self, app_handle: &AppHandle, project: StoredJavaProject) -> AppResult<()> {
        let mut storage = self.load_projects(app_handle).await?;

        if let Some(existing) = storage.projects.iter_mut().find(|p| p.id == project.id) {
            *existing = project;
        }

        self.save_projects(app_handle, &storage).await
    }

    /// Set project open state
    pub async fn set_project_open(&self, app_handle: &AppHandle, project_id: &str, is_open: bool) -> AppResult<()> {
        let mut storage = self.load_projects(app_handle).await?;

        if let Some(project) = storage.projects.iter_mut().find(|p| p.id == project_id) {
            project.is_open = is_open;
        }

        self.save_projects(app_handle, &storage).await
    }

    /// Delete a Java project reference
    pub async fn delete_project(&self, app_handle: &AppHandle, project_id: &str) -> AppResult<()> {
        let mut storage = self.load_projects(app_handle).await?;
        storage.projects.retain(|p| p.id != project_id);
        self.save_projects(app_handle, &storage).await
    }

    /// Mark endpoints as seen for a project
    pub async fn mark_endpoints_seen(&self, app_handle: &AppHandle, project_id: &str, endpoint_ids: Vec<String>) -> AppResult<()> {
        let mut storage = self.load_projects(app_handle).await?;

        if let Some(project) = storage.projects.iter_mut().find(|p| p.id == project_id) {
            // Merge new endpoint IDs with existing ones
            for id in endpoint_ids {
                if !project.seen_endpoint_ids.contains(&id) {
                    project.seen_endpoint_ids.push(id);
                }
            }
        }

        self.save_projects(app_handle, &storage).await
    }

    /// Get new (unseen) endpoints by comparing parsed data with seen IDs
    pub fn get_new_endpoints(&self, project: &StoredJavaProject, parsed_data: &ParsedJavaProject) -> Vec<String> {
        let mut new_endpoint_ids = Vec::new();

        for controller in &parsed_data.controllers {
            for endpoint in &controller.endpoints {
                // An endpoint is "new" if its ID is not in the seen list
                // We use a composite key: controller_name + method + path
                let composite_key = format!("{}:{}:{}", controller.name, endpoint.http_method, endpoint.full_path);
                if !project.seen_endpoint_ids.contains(&composite_key) {
                    new_endpoint_ids.push(endpoint.id.clone());
                }
            }
        }

        new_endpoint_ids
    }
}

impl Default for JavaProjectStore {
    fn default() -> Self {
        Self::new()
    }
}
